code/Forms/UnsavedChangesIndicator.php

```diff
@@ -0,0 +1,66 @@
+<?php
+
+namespace SilverStripe\Admin\Forms;
+
+use RuntimeException;
+use SilverStripe\Forms\FormAction;
+
+/**
+ * A visual indicator to show whether a form has unsaved changes.
+ *
+ * Note that this isn't strictly a FormAction as it doesn't perform any action when clicked.
+ * However, extending FormAction allows us to easily integrate with the existing form action
+ * architecture (e.g. placement in the form actions area).
+ */
+class UnsavedChangesIndicator extends FormAction
+{
+    private static array $minutes = [
+        'notice' => 5,
+        'warning' => 10,
+    ];
+
+    protected $schemaComponent = 'UnsavedChangesIndicator';
+
+    public function getSchemaDataDefaults()
+    {
+        // This is used for initialising the React component
+        // Refer to silverstrpie/admin UnsavedChangesIndicator.js
+        $data = parent::getSchemaDataDefaults();
+        $data['minutes'] = $this->getMinutes();
+        return $data;
+    }
+
+    public function getAttributes()
+    {
+        // This is used for entwine initialisation
+        // Refer to silverstrpie/admin UnsavedChangesIndicatorEntwine.js
+        $attributes = parent::getAttributes();
+        $attributes['data-minutes'] = json_encode($this->getMinutes());
+        return $attributes;
+    }
+
+    private function getMinutes(): array
+    {
+        $minutes = UnsavedChangesIndicator::config()->get('minutes');
+        $this->validateConfig($minutes);
+        return $minutes;
+    }
+
+    private function validateConfig(array $minutes): void
+    {
+        if (!isset($minutes['notice']) || !isset($minutes['warning'])) {
```

Comment:
```suggestion
        if (!array_key_exists('notice', $minutes) || !array_key_exists('warning', $minutes)) {
```
This is a bit of a nitpick - but technically with a value of `null` for either of these keys we should hit the "Config values must be numeric" exception since the keys _are_ present, just not with valid values.

It's not that important though - I won't block merging if this isn't changed.

Reply:
Updated